prov/sm2/src/sm2_progress.c

```diff
@@ -166,47 +166,63 @@ static int sm2_progress_recv_msg(struct sm2_ep *ep,
 	return ret < 0 ? ret : 0;
 }
 
-void sm2_progress_recv(struct sm2_ep *ep)
+int sm2_progress_proto_return(struct sm2_ep *ep,
+			      struct sm2_xfer_entry *xfer_entry)
 {
+	int ret;
 	struct sm2_av *av =
 		container_of(ep->util_ep.av, struct sm2_av, util_av);
 	struct sm2_mmap *map = &av->mmap;
+
+	if (xfer_entry->hdr.op_flags & FI_DELIVERY_COMPLETE) {
+		ret = sm2_complete_tx(ep, (void *) xfer_entry->hdr.context,
+				      xfer_entry->hdr.op,
+				      xfer_entry->hdr.op_flags);
+		if (ret)
+			FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+				"Unable to process "
+				"FI_DELIVERY_COMPLETE "
+				"completion\n");
+	}
+
+	smr_freestack_push(sm2_freestack(sm2_mmap_ep_region(map, ep->gid)),
+			   xfer_entry);
+	return ret;
+}
+
+int sm2_progress_proto_inject(struct sm2_ep *ep,
+			      struct sm2_xfer_entry *xfer_entry)
+{
+	int ret;
+	switch (xfer_entry->hdr.op) {
+	case ofi_op_msg:
+	case ofi_op_tagged:
+		ret = sm2_progress_recv_msg(ep, xfer_entry);
+		break;
+	default:
+		FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+			"Unexpected op for inject proto\n");
+		ret = -FI_EINVAL;
+	}
+	return ret;
+}
+
+void sm2_progress_recv(struct sm2_ep *ep)
+{
 	struct sm2_xfer_entry *xfer_entry;
 	int ret = 0, i;
 
 	for (i = 0; i < MAX_SM2_MSGS_PROGRESSED; i++) {
 		xfer_entry = sm2_fifo_read(ep);
 		if (!xfer_entry)
 			break;
-
-		if (xfer_entry->hdr.proto == sm2_proto_return) {
-			if (xfer_entry->hdr.op_flags & FI_DELIVERY_COMPLETE) {
-				ret = sm2_complete_tx(
-					ep, (void *) xfer_entry->hdr.context,
-					xfer_entry->hdr.op,
-					xfer_entry->hdr.op_flags);
-				if (ret)
-					FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
-						"Unable to process "
-						"FI_DELIVERY_COMPLETE "
-						"completion\n");
-			}
-
-			smr_freestack_push(
-				sm2_freestack(sm2_mmap_ep_region(map, ep->gid)),
-				xfer_entry);
-			continue;
-		}
-
-		switch (xfer_entry->hdr.op) {
-		case ofi_op_msg:
-		case ofi_op_tagged:
-			ret = sm2_progress_recv_msg(ep, xfer_entry);
+		switch (xfer_entry->hdr.proto) {
+		case (sm2_proto_return):
```

Comment:
Case 1: Progress
```
sm2_progress_recv(): Switch 1 PROTO 
sm2_progress_proto_inject(): Switch 2 OP
sm2_start_common(): Switch 3 PROTO     #### BAD, we shouldn't switch twice on the same thing... this needs to go
```

Case 2 posting buffers:
```
sm2_start_common(): Switch 1 PROTO  # this switch needs to happen on unexpected msg path
```

I think part of this change needs to get rid of sm2_start_common() b/c now those paths are diverging.